auto mission fly home before complete loss of global position estimate

dagar opened the pull request:
In a plane with airspeed the default estimator (ekf2) is able to continue providing a usable global position estimate after the loss of GPS. Currently in master if you lose GPS the system continues operating normally until the error of the position estimate exceeds the configured thresholds, then a failsafe triggers.

This pull request updates the reporting and failsafe behavior so that we use the dead reckoning period to try and return home.

 - ekf2: expose dead reckoning as control status flags
 - commander:
    - add GPS validity check
    - in AUTO MISSION if dependent on GPS then a loss of GPS will
trigger RTL so that the plane can return home while dead reckoning

#### Notes
 - currently only applies to AUTO_MISSION, but probably makes sense in other places if the behavior is configurable

dagar commented on msg/estimator_status_flags.msg:
```diff
@@ -32,6 +32,8 @@ bool cs_ev_vel                # 24 - true when local frame velocity data fusion
 bool cs_synthetic_mag_z       # 25 - true when we are using a synthesized measurement for the magnetometer Z component
 bool cs_vehicle_at_rest       # 26 - true when the vehicle is at rest
 bool cs_gps_yaw_fault         # 27 - true when the GNSS heading has been declared faulty and is no longer being used
+bool cs_inertial_dead_reckoning # 28 - true if we are no longer fusing measurements that constrain horizontal velocity drift
```
@bresch do these make sense as control status flags?

dagar commented on src/modules/commander/Commander.cpp:
```diff
@@ -3784,51 +3786,71 @@ void Commander::estimator_check()
+			bool optical_flow    = estimator_status_flags.cs_opt_flow;
+			bool vision_position = estimator_status_flags.cs_ev_pos;
+
+			_status_flags.position_reliant_on_gps             =  gps && !optical_flow && !vision_position;
```
Is there a better way to latch onto which position sources we require?

dagar replied:
For example we only ever had GPS, if it disappears in flight in flight `estimator_status_flags.cs_gps` will no longer be true. How do we know `GPS` is required?

bresch replied:
The information regarding what's required is not stored anywhere AFAIK. EKF2 uses the data it gets if allowed by the aidmask param and that's all. I think your proposal here makes sense.

bresch commented on src/modules/commander/Commander.cpp:
```diff
@@ -3784,51 +3786,71 @@ void Commander::estimator_check()
-		const bool mag_fault = (estimator_status.control_mode_flags & (1 << estimator_status_s::CS_MAG_FAULT));
-		const bool gnss_heading_fault = (estimator_status.control_mode_flags & (1 << estimator_status_s::CS_GPS_YAW_FAULT));
+		if (estimator_status_flags.cs_inertial_dead_reckoning || estimator_status_flags.cs_wind_dead_reckoning) {
+			_status_flags.position_dead_reckoning = true;
```
"Inertial dead reckoning" means: nav solution obtained by previous solution and inertial measurements. Therefore, "Position dead reckoning" sounds a bit strange to me. If in inertial or wind DR, why not calling it simply "dead_reckoning"?

dagar replied:
In the context of `vehicle_status_flags` I was trying to be more descriptive. I'll update it.

bresch commented on src/modules/ekf2/EKF2.cpp:
```diff
@@ -1229,7 +1230,8 @@ void EKF2::PublishStatusFlags(const hrt_abstime &timestamp)
 		status_flags.cs_ev_vel                = _ekf.control_status_flags().ev_vel;
 		status_flags.cs_synthetic_mag_z       = _ekf.control_status_flags().synthetic_mag_z;
 		status_flags.cs_vehicle_at_rest       = _ekf.control_status_flags().vehicle_at_rest;
-		status_flags.cs_gps_yaw_fault         = _ekf.control_status_flags().gps_yaw_fault;
```
I agree that the `gps_yaw_fault` flag should go in the fault status flags, but you shouldn't just remove it :D

dagar replied:
Thanks, fixed.

ryanjAA commented:
@dagar this is awesome!

sfuhrer commented:
Yep that's an important one. I think it should be handled for all Auto modes the same way, and we need to think about which settings make sense. I would argue that it is something that's not "vehicle"-specific but rather "mission"-specific, meaning that it needs a setting that is exposed in the Safety tab of QGC rather than just being a mere parameter. Something like:
GPS Loss Failsafe Trigger
- 0: only warn user (proceeds flight/mission normally as long as the position estimate is valid via dead reckoning)
- 1: RTL
We should also think about when we actually want to declare the position estimate invalid in the dead reckoning case. Can we just increase COM_POS_FS_EPH and COM_POS_FS_EPV? In most cases a RTL with even a very bad position accuracy is better then the "Descend with fixed-bank loiter at current location" logic that is triggered if the position estimate is declared invalid.

dagar commented:
> We should also think about when we actually want to declare the position estimate invalid in the dead reckoning case. Can we just increase COM_POS_FS_EPH and COM_POS_FS_EPV? In most cases a RTL with even a very bad position accuracy is better then the "Descend with fixed-bank loiter at current location" logic that is triggered if the position estimate is declared invalid.

Yes definitely. It's already a little weird because those existing thresholds are actually 2.5x if already valid. 
https://github.com/PX4/PX4-Autopilot/blob/8b1543e3b06de466fc89b7b3ec5baf33cbf876ee/src/modules/commander/Commander.cpp#L3121

We also have different defaults for FW vs MC, but VTOL doesn't get the benefit when operating as FW.
https://github.com/PX4/PX4-Autopilot/blob/8b1543e3b06de466fc89b7b3ec5baf33cbf876ee/ROMFS/px4fmu_common/init.d/rc.fw_defaults#L14-L15

The idea was to eventually make that contextual, if you're up and away flying as a plane we can raise the threshold significantly, but then we have to think about what happens after you get home and want to actually land or transition back to a multicopter.

This might be a good one to revisit in detail once the flight modes are fully consolidated and isolated per vehicle type.

sfuhrer commented:
>const float req_accuracy = (was_valid ? required_accuracy * 2.5f : required_accuracy); 

Anything against raising this factor to around 5 after the logic of this PR is in? Or do we need to keep it so tight for MC? How bad would it be to do a distinction based on `vtol_in_rw_mode` here?

>We also have different defaults for FW vs MC, but VTOL doesn't get the benefit when operating as FW.

I've raised the thresholds for our VTOLs to about the FW values. Don't see a big downside, given that VTOLs are at least 90% in FW mode and for take-off/landings I don't see why switching out of a Position controlled flight to a failsafe (aka descend) can be any safer. 

>The idea was to eventually make that contextual, if you're up and away flying as a plane we can raise the threshold significantly, but then we have to think about what happens after you get home and want to actually land or transition back to a multicopter.

Yes totally. Let's though also find something for the mean time. See my above points. Happy to open another PR for it, as this one here is kind of decoupled.

dagar commented:
> > const float req_accuracy = (was_valid ? required_accuracy * 2.5f : required_accuracy);
> 
> Anything against raising this factor to around 5 after the logic of this PR is in? Or do we need to keep it so tight for MC? How bad would it be to do a distinction based on `vtol_in_rw_mode` here?
> 
> > We also have different defaults for FW vs MC, but VTOL doesn't get the benefit when operating as FW.
> 
> I've raised the thresholds for our VTOLs to about the FW values. Don't see a big downside, given that VTOLs are at least 90% in FW mode and for take-off/landings I don't see why switching out of a Position controlled flight to a failsafe (aka descend) can be any safer.
> 
> > The idea was to eventually make that contextual, if you're up and away flying as a plane we can raise the threshold significantly, but then we have to think about what happens after you get home and want to actually land or transition back to a multicopter.
> 
> Yes totally. Let's though also find something for the mean time. See my above points. Happy to open another PR for it, as this one here is kind of decoupled.

The increased multiplier is probably fine, although it might start to get questionable if someone already knew enough that they were adjusting `COM_POS_FS_EPH` to account for FW loss of GPS.

I think I'd like to get a better understanding of the underlying requirements and history here. For example can we generally have a much more relaxed `COM_POS_FS_EPH` other than when it matters in the context of a mission (or similar)? Maybe for multi-copter it's primarily `COM_VEL_FS_EVH` we need to keep an eye on, and then absolute position accuracy can be much more contextual.

dagar commented:
Comments addressed and rebased on master (minor merge conflicts resolved). This has been tested in SITL.

For now I've updated `vehicle_status_flags.dead_reckoning` to only use cs_wind_dead_reckoning and avoid uselessly triggering RTL for a few seconds before a full position loss (if airspeed isn't available).

Then the only remaining hesitation I have at this stage is if the simple GPS validity check is appropriate for real world data and not going to have any false positives.

Antiheavy commented:
I think adding a GPS Validity flag is a huge improvement for multiple reasons.  Giving the user a warning of a potential impending loss of global position failsafe is really important.  It can also help bring to the surface intermittent GPS/GNSS problems that are masked by the current architecture.

FWIW -  I'm skeptical that dead reckoning is accurate enough to perform complex flight guidance required for a fixed wing RTL. we've tried to execute a fixed wing RTL (return to landing pattern) multiple times and end up being many 100's of meters, even 1000+meters off true position on final approach.  Our FW system is fairly simple: FMUv4, gps, airspeed, external mag.   We are on a pretty old PX4 branch, but EKF2 hasn't changed that much, so take that for what its worth as well.  We've found that in a stable loiter the system will track pretty well for several minutes, but more complicated flight maneuvers cause dead reckoning to diverge quickly from reality.  The systems EPH estimates seem way to optimistic.  I suspect non-perfectly calibrated airspeed measurements (very typical in small UAS) contribute to dead reckoning error.  I wish I had logs handy to share, Flight Review deletes them too quickly....

Setting aside my opinion above, I suppose there is always an argument to "at least try" to RTL.  Personally I'd probably leave the parameter option to "continue mission" and hope that GPS comes back.  If it doesn't come back then I'd rely on the "loiter down" fail-safe behavior before the aircraft flies away too far to be able to go find.
